Rocksdb Error when Downgrading from Besu 23.7.1 to 22.10.2

# Rocksdb Error when Downgrading from Besu 23.7.1 to 22.10.2

- [ x] Have you read the Code of Conduct? By filing an Issue, I agree to comply with it, including treating everyone with respect: [Code of Conduct]()
- [ x] Reproduced the issue in the latest version of the software
- [ x] Read the debugging docs: [Besu Debugging Docs]()
- [ x] Checked for duplicate issues: [Besu Issues]()

## Description
I want to seamlessly revert from version 23.7.1 to 22.10.2 without any database errors after having successfully upgraded to 23.7.1 and executed the recommended database revert command, and finally downgrading back to 22.10.2.

## Acceptance Criteria
Besu should revert back to version 22.10.2 without any startup errors after previously upgrading to 23.7.1 and executing the recommended database revert commands. See [here](https://github.com/hyperledger/besu/pull/5471) as recommended.

## Steps to Reproduce (Bug)
1. Deploy Besu 22.10.2 (with 4 min validators) as a private network.
2. Allow blocks to be imported.
3. Stop Besu 22.10.2 and run besu with version 23.7.1.
4. Allow it to import blocks on 23.7.1.
5. Stop the 23.7.1 version and execute the command: `besu revert storage-variables`
6. Redeploy using version 22.10.2. An error should be encountered during startup.

**Expected behavior:** Besu 22.10.2 should start without errors after reverting from 23.7.1.

**Actual behavior:** Encountered an error during the startup of Besu 22.10.2 after reverting.

**Frequency:** This issue occurs consistently upon reverting.

### Logs when reverting db via executing subcommand `storage revert-variables`
```
2023-08-18 14:29:43.219+00:00 | main | INFO | DatabaseMetadata | Lookup database metadata file in data directory: /besu/data
2023-08-18 14:29:43.835+00:00 | main | INFO | RocksDBKeyValueStorageFactory | Existing database detected at /besu/data. Version 1. Compacting database..
2023-08-18 14:29:52.733+00:00 | main | INFO | StorageSubCommand$Revert VariablesStorage | Reverted variable storage for key chainHeadHash
2023-08-18 14:29:52.821+00:00 | main | INFO | StorageSubCommand$RevertVariablesStorage | Reverted variable storage for key local-enr-seqno
```


### Logs (if a bug)
```
2023-08-22 00:11:15.041+00:00 I main I INFO I RocksDBKeyValueStorageFactory | Existing database detected at /besu/data. Version 1
picocli.CommandLine$ParameterException: org.rocksdb.RocksDBException: Column families not opened: O
at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1443)
at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
at picocli.CommandLine.access$1300(CommandLine.java:145)
at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
at picocli.CommandLine.execute(CommandLine.java:2078)
at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:54)
at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:32)
at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
at picocli.CommandLine.execute(CommandLine.java:2078)
at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1614)
at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1391)
at org.hyperledger.besu.Besu.main(Besu.java:48)
Caused by: picocli.CommandLine$ExecutionException: org.rocksdb.RocksDBException: Column families not opened: O
at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:2122)
at org.hyperledger.besu.cli.BesuCommand.initController(BesuCommand.java:2115)
at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1427)
... 16 more
Caused by: org.hyperledger.besu.plugin.services.exception.StorageException: org.rocksdb.RocksDBException: Column families not opened: D
at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage.<init>(RocksDBColumnarKeyValueStorage.java:166)
at org.hyperledger.besu.plugin.services.storage.rocksdb.RocksDBKeyValueStorageFactory.create(RocksDBKeyValueStorageFactory.java:124)
at org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStorageProviderBuilder.build(KeyValueStorageProviderBuilder.java:71)
at org.hyperledger.besu.cli.BesuCommand.keyValueStorageProvider(BesuCommand.java:2803)
at org.hyperledger.besu.cli.BesuCommand.getControllerBuilder(BesuCommand.java:2127)
at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:2120)
Caused by: org.rocksdb.RocksDBException: Column families not opened: [
at org.rocksdb.OptimisticTransactionDB.open(Native Method)
at org.rocksdb.OptimisticTransactionDB.open(OptimisticTransactionDB.java:80)
at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage.<init>(RocksDBColumnarKeyValueStorage.java:146)
... 23 more
To display full help: besu [COMMAND] --help

```

#### Software version: Initially besu 22.10.2 then upgraded to 23.7.1. Finally wanted to revert back to 22.10.2
#### Additional Information
#### Database storage format : FOREST

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,19 @@
 # Changelog
 
+## 23.7.3
+### Additions and Improvements
+
+### Breaking Changes
+- Removed support for Kotti network (ETC) [#5816](https://github.com/hyperledger/besu/pull/5816)
+
+### Additions and Improvements
+
+### Bug Fixes
+- do not create ignorable storage on revert storage-variables subcommand [#5830](https://github.com/hyperledger/besu/pull/5830) 
+
+### Download Links
+
+
 ## 23.7.2
 
 ### Additions and Improvements
@@ -30,7 +44,6 @@
 ### Breaking Changes
 - Removed deprecated GoQuorum permissioning interop [#5607](https://github.com/hyperledger/besu/pull/5607)
 - Removed support for version 0 of the database as it is no longer used by any active node. [#5698](https://github.com/hyperledger/besu/pull/5698)
-- Removed support for Kotti network (ETC) [#5816](https://github.com/hyperledger/besu/pull/5816)
 
 ### Additions and Improvements
 - `evmtool` launcher binaries now ship as part of the standard distribution. [#5701](https://github.com/hyperledger/besu/pull/5701)

--- a/besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java
+++ b/besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java
@@ -3489,6 +3489,7 @@ public class BesuCommand implements DefaultCommandValues, Runnable {
         getActualGenesisConfigOptions().getTerminalTotalDifficulty().isPresent());
   }
 
+  /** Set ignorable segments in RocksDB Storage Provider plugin. */
   public void setIgnorableStorageSegments() {
     if (!unstableChainPruningOptions.getChainDataPruningEnabled()) {
       rocksDBPlugin.addIgnorableSegmentIdentifier(KeyValueSegmentIdentifier.CHAIN_PRUNER_STATE);

--- a/besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java
+++ b/besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java
@@ -3489,7 +3489,7 @@ public class BesuCommand implements DefaultCommandValues, Runnable {
         getActualGenesisConfigOptions().getTerminalTotalDifficulty().isPresent());
   }
 
-  private void setIgnorableStorageSegments() {
+  public void setIgnorableStorageSegments() {
     if (!unstableChainPruningOptions.getChainDataPruningEnabled()) {
       rocksDBPlugin.addIgnorableSegmentIdentifier(KeyValueSegmentIdentifier.CHAIN_PRUNER_STATE);
     }

--- a/besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/StorageSubCommand.java
+++ b/besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/StorageSubCommand.java
@@ -99,6 +99,8 @@ public class StorageSubCommand implements Runnable {
     }
 
     private StorageProvider getStorageProvider() {
+      // init collection of ignorable segments
+      parentCommand.parentCommand.setIgnorableStorageSegments();
       return parentCommand.parentCommand.getStorageProvider();
     }
 

--- a/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/OptimisticRocksDBColumnarKeyValueStorage.java
+++ b/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/OptimisticRocksDBColumnarKeyValueStorage.java
@@ -63,7 +63,7 @@ public class OptimisticRocksDBColumnarKeyValueStorage extends RocksDBColumnarKey
       initColumnHandles();
 
     } catch (final RocksDBException e) {
-      throw new StorageException(e);
+      throw parseRocksDBException(e, segments, ignorableSegments);
     }
   }
 

--- a/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java
+++ b/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java
@@ -168,6 +168,14 @@ public abstract class RocksDBColumnarKeyValueStorage implements SegmentedKeyValu
     }
   }
 
+  /**
+   * Parse RocksDBException and wrap in StorageException
+   *
+   * @param ex RocksDBException
+   * @param defaultSegments segments requested to open
+   * @param ignorableSegments segments which are ignorable if not present
+   * @return StorageException wrapping the RocksDB Exception
+   */
   protected static StorageException parseRocksDBException(
       final RocksDBException ex,
       final List<SegmentIdentifier> defaultSegments,

--- a/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java
+++ b/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java
@@ -16,6 +16,9 @@ package org.hyperledger.besu.plugin.services.storage.rocksdb.segmented;
 
 import static java.util.stream.Collectors.toUnmodifiableSet;
 
+import com.google.common.base.Splitter;
+import com.google.common.collect.Streams;
+import org.apache.tuweni.bytes.Bytes;
 import org.hyperledger.besu.plugin.services.MetricsSystem;
 import org.hyperledger.besu.plugin.services.exception.StorageException;
 import org.hyperledger.besu.plugin.services.metrics.OperationTimer;
@@ -31,6 +34,7 @@ import org.hyperledger.besu.plugin.services.storage.rocksdb.configuration.RocksD
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -161,7 +165,42 @@ public abstract class RocksDBColumnarKeyValueStorage implements SegmentedKeyValu
       txOptions = new TransactionDBOptions();
       columnHandles = new ArrayList<>(columnDescriptors.size());
     } catch (RocksDBException e) {
-      throw new StorageException(e);
+      List<SegmentIdentifier> knownSegments = Streams.concat(
+              defaultSegments.stream(),
+              ignorableSegments.stream())
+          .distinct()
+          .toList();
+      throw parseRocksDBException(e, knownSegments);
+    }
+  }
+
+  private static StorageException parseRocksDBException(
+      final RocksDBException ex,
+      final List<SegmentIdentifier> knownSegments) {
+    String message = ex.getMessage();
+
+    // parse out unprintable segment names for a more useful exception:
+    String columnExceptionMessagePrefix = "Column families not opened: ";
+    if (message.contains(columnExceptionMessagePrefix)) {
+      String substring = message.substring(message.indexOf(": ") + 2);
+
+      List<String> unHandledSegments = new ArrayList<>();
+      Splitter.on(", ").splitToStream(substring)
+          .forEach(part -> {
+            byte[] bytes = part.getBytes(StandardCharsets.UTF_8);
+            unHandledSegments.add(
+            knownSegments.stream().filter(seg -> Arrays.equals(seg.getId(), bytes))
+                .findFirst()
+                .map(SegmentIdentifier::getName)
+                .orElse("unknown segment:{" + Bytes.of(bytes).toHexString() + "}"));
+
+          });
+
+      return new StorageException(
+          "RocksDBException: Unhandled column families: [" +
+              unHandledSegments.stream().collect(Collectors.joining(", ")) + "]");
+    } else {
+      return new StorageException(ex);
     }
   }
 

--- a/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java
+++ b/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java
@@ -16,9 +16,6 @@ package org.hyperledger.besu.plugin.services.storage.rocksdb.segmented;
 
 import static java.util.stream.Collectors.toUnmodifiableSet;
 
-import com.google.common.base.Splitter;
-import com.google.common.collect.Streams;
-import org.apache.tuweni.bytes.Bytes;
 import org.hyperledger.besu.plugin.services.MetricsSystem;
 import org.hyperledger.besu.plugin.services.exception.StorageException;
 import org.hyperledger.besu.plugin.services.metrics.OperationTimer;
@@ -34,7 +31,6 @@ import org.hyperledger.besu.plugin.services.storage.rocksdb.configuration.RocksD
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -44,7 +40,10 @@ import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import com.google.common.base.Splitter;
+import com.google.common.collect.Streams;
 import org.apache.commons.lang3.tuple.Pair;
+import org.apache.tuweni.bytes.Bytes;
 import org.rocksdb.BlockBasedTableConfig;
 import org.rocksdb.BloomFilter;
 import org.rocksdb.ColumnFamilyDescriptor;
@@ -165,18 +164,14 @@ public abstract class RocksDBColumnarKeyValueStorage implements SegmentedKeyValu
       txOptions = new TransactionDBOptions();
       columnHandles = new ArrayList<>(columnDescriptors.size());
     } catch (RocksDBException e) {
-      List<SegmentIdentifier> knownSegments = Streams.concat(
-              defaultSegments.stream(),
-              ignorableSegments.stream())
-          .distinct()
-          .toList();
+      List<SegmentIdentifier> knownSegments =
+          Streams.concat(defaultSegments.stream(), ignorableSegments.stream()).distinct().toList();
       throw parseRocksDBException(e, knownSegments);
     }
   }
 
   private static StorageException parseRocksDBException(
-      final RocksDBException ex,
-      final List<SegmentIdentifier> knownSegments) {
+      final RocksDBException ex, final List<SegmentIdentifier> knownSegments) {
     String message = ex.getMessage();
 
     // parse out unprintable segment names for a more useful exception:
@@ -185,20 +180,23 @@ public abstract class RocksDBColumnarKeyValueStorage implements SegmentedKeyValu
       String substring = message.substring(message.indexOf(": ") + 2);
 
       List<String> unHandledSegments = new ArrayList<>();
-      Splitter.on(", ").splitToStream(substring)
-          .forEach(part -> {
-            byte[] bytes = part.getBytes(StandardCharsets.UTF_8);
-            unHandledSegments.add(
-            knownSegments.stream().filter(seg -> Arrays.equals(seg.getId(), bytes))
-                .findFirst()
-                .map(SegmentIdentifier::getName)
-                .orElse("unknown segment:{" + Bytes.of(bytes).toHexString() + "}"));
-
-          });
+      Splitter.on(", ")
+          .splitToStream(substring)
+          .forEach(
+              part -> {
+                byte[] bytes = part.getBytes(StandardCharsets.UTF_8);
+                unHandledSegments.add(
+                    knownSegments.stream()
+                        .filter(seg -> Arrays.equals(seg.getId(), bytes))
+                        .findFirst()
+                        .map(SegmentIdentifier::getName)
+                        .orElse("unknown segment:{" + Bytes.of(bytes).toHexString() + "}"));
+              });
 
       return new StorageException(
-          "RocksDBException: Unhandled column families: [" +
-              unHandledSegments.stream().collect(Collectors.joining(", ")) + "]");
+          "RocksDBException: Unhandled column families: ["
+              + unHandledSegments.stream().collect(Collectors.joining(", "))
+              + "]");
     } else {
       return new StorageException(ex);
     }

--- a/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java
+++ b/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java
@@ -164,15 +164,17 @@ public abstract class RocksDBColumnarKeyValueStorage implements SegmentedKeyValu
       txOptions = new TransactionDBOptions();
       columnHandles = new ArrayList<>(columnDescriptors.size());
     } catch (RocksDBException e) {
-      List<SegmentIdentifier> knownSegments =
-          Streams.concat(defaultSegments.stream(), ignorableSegments.stream()).distinct().toList();
-      throw parseRocksDBException(e, knownSegments);
+      throw parseRocksDBException(e, defaultSegments, ignorableSegments);
     }
   }
 
-  private static StorageException parseRocksDBException(
-      final RocksDBException ex, final List<SegmentIdentifier> knownSegments) {
+  protected static StorageException parseRocksDBException(
+      final RocksDBException ex,
+      final List<SegmentIdentifier> defaultSegments,
+      final List<SegmentIdentifier> ignorableSegments) {
     String message = ex.getMessage();
+    List<SegmentIdentifier> knownSegments =
+        Streams.concat(defaultSegments.stream(), ignorableSegments.stream()).distinct().toList();
 
     // parse out unprintable segment names for a more useful exception:
     String columnExceptionMessagePrefix = "Column families not opened: ";
@@ -189,8 +191,9 @@ public abstract class RocksDBColumnarKeyValueStorage implements SegmentedKeyValu
                     knownSegments.stream()
                         .filter(seg -> Arrays.equals(seg.getId(), bytes))
                         .findFirst()
-                        .map(SegmentIdentifier::getName)
-                        .orElse("unknown segment:{" + Bytes.of(bytes).toHexString() + "}"));
+                        .map(seg -> new SegmentRecord(seg.getName(), seg.getId()))
+                        .orElse(new SegmentRecord(part, bytes))
+                        .forDisplay());
               });
 
       return new StorageException(
@@ -393,4 +396,10 @@ public abstract class RocksDBColumnarKeyValueStorage implements SegmentedKeyValu
   }
 
   abstract RocksDB getDB();
+
+  record SegmentRecord(String name, byte[] id) {
+    public String forDisplay() {
+      return String.format("'%s'(%s)", name, Bytes.of(id).toHexString());
+    }
+  }
 }

--- a/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/TransactionDBRocksDBColumnarKeyValueStorage.java
+++ b/plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/TransactionDBRocksDBColumnarKeyValueStorage.java
@@ -66,7 +66,7 @@ public class TransactionDBRocksDBColumnarKeyValueStorage extends RocksDBColumnar
       initColumnHandles();
 
     } catch (final RocksDBException e) {
-      throw new StorageException(e);
+      throw parseRocksDBException(e, segments, ignorableSegments);
     }
   }
 

--- a/plugins/rocksdb/src/test/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorageTest.java
+++ b/plugins/rocksdb/src/test/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorageTest.java
@@ -222,7 +222,7 @@ public abstract class RocksDBColumnarKeyValueStorageTest extends AbstractKeyValu
       createSegmentedStore(testPath, Arrays.asList(TestSegment.FOO, TestSegment.BAR), List.of());
       fail("DB without knowledge of experimental column family should fail");
     } catch (StorageException e) {
-      assertThat(e.getMessage()).contains("Column families not opened");
+      assertThat(e.getMessage()).contains("Unhandled column families");
     }
 
     // Even if the column family is marked as ignored, as long as it exists, it will not be ignored
@@ -265,7 +265,7 @@ public abstract class RocksDBColumnarKeyValueStorageTest extends AbstractKeyValu
       createSegmentedStore(testPath, Arrays.asList(TestSegment.FOO, TestSegment.BAR), List.of());
       fail("DB without knowledge of experimental column family should fail");
     } catch (StorageException e) {
-      assertThat(e.getMessage()).contains("Column families not opened");
+      assertThat(e.getMessage()).contains("Unhandled column families");
     }
   }
 

